File: mediator/typescript/Mediator.ts
// Mediator interface
interface ChatMediator {
  sendMessage(message: string, user: User): void;
  addUser(user: User): void;
}

// Concrete Mediator
class ChatRoom implements ChatMediator {
  private users: User[] = [];

  addUser(user: User): void {
    this.users.push(user);
  }

  sendMessage(message: string, sender: User): void {
    this.users.forEach(user => {
      if (user !== sender) {
        user.receive(message, sender.getName());
      }
    });
  }
}

// User
class User {
  private name: string;
  private mediator: ChatMediator;
  private messages: string[] = [];

  constructor(name: string, mediator: ChatMediator) {
    this.name = name;
    this.mediator = mediator;
    mediator.addUser(this);
  }

  getName(): string {
    return this.name;
  }

  send(message: string): void {
    console.log(`${this.name} sends: ${message}`);
    this.mediator.sendMessage(message, this);
  }

  receive(message: string, from: string): void {
    const msg = `${this.name} receives from ${from}: ${message}`;
    this.messages.push(msg);
    console.log(msg);
  }
}

// Usage
const chatRoom = new ChatRoom();
const alice = new User('Alice', chatRoom);
const bob = new User('Bob', chatRoom);
const charlie = new User('Charlie', chatRoom);

alice.send('Hello everyone!');
bob.send('Hi Alice!');

File: patterns/mediator/python/Mediator.py
from __future__ import annotations
from abc import ABC, abstractmethod
from typing import List

# Mediator interface
class ChatMediator(ABC):
    @abstractmethod
    def send_message(self, message: str, user: "User"):
        pass

    @abstractmethod
    def add_user(self, user: "User"):
        pass

# Concrete Mediator
class ChatRoom(ChatMediator):
    def __init__(self):
        self._users: List[User] = []

    def add_user(self, user: "User"):
        self._users.append(user)

    def send_message(self, message: str, sender: "User"):
        for user in self._users:
            if user is not sender:
                user.receive(message, sender.name)

# User
class User:
    def __init__(self, name: str, mediator: ChatMediator):
        self.name = name
        self._mediator = mediator
        self._messages: List[str] = []
        mediator.add_user(self)

    def send(self, message: str):
        print(f"{self.name} sends: {message}")
        self._mediator.send_message(message, self)

    def receive(self, message: str, from_user: str):
        msg = f"{self.name} receives from {from_user}: {message}"
        self._messages.append(msg)
        print(msg)


if __name__ == "__main__":
    chat_room = ChatRoom()

    alice = User("Alice", chat_room)
    bob = User("Bob", chat_room)
    charlie = User("Charlie", chat_room)

    alice.send("Hello everyone!")
    bob.send("Hi Alice!")
    charlie.send("Hey folks!")
